### DISCUSSION TITLE
Maximum Flow and Minimum-Cost Flow in Almost-Linear Time

### DISCUSSION THREAD

**Discussion by ODAncona** [networkx/networkx#5716] [Ideas]
A research team found a way to solve max flow problem in **quasi linear time** using customized data structure. They published their work in April 2022.

> It computes exact maximum flows and minimum-cost flows on
directed graphs with m edges and polynomially bounded integral demands, costs, and capacities
in $m^{1+o(1)}$ time. Our algorithm builds the flow through a sequence of  $m^{1+o(1)}$ approximate
undirected minimum-ratio cycles, each of which is computed and processed in amortized $m^{o(1)}$
time using a new dynamic graph data structure.

[Press article](https://www.quantamagazine.org/researchers-achieve-absurdly-fast-algorithm-for-network-flow-20220608/)
[Paper](https://arxiv.org/abs/2203.00671)

I wonder if it's possible to integrate this algorithm to networkx somehow?

In case you didn't knew it was possible until someone did it!

**Comment by dschult** [comment]
This would definitely be of interest.  I haven’t looked at the paper, but you need to be careful with o(1) powers of m if we don’t have any indication of what that power is. Still, NetworkX might be a good place to try the implementation and see what that power is (or maybe the article already discusses this).  In any case, we are interested for sure.

**Reply by ODAncona** [reply]
In the article they said,

> For now, it’s primarily a theoretical advance, since the speed improvements kick in only for networks that are far larger than the ones we encounter in the real world, for which maximum flow problems can already be solved fairly quickly (at least, if they don’t involve minimizing costs). But pieces of the new algorithm might see practical use within a year, predicted [Richard Peng](https://cs.uwaterloo.ca/people-profiles/richard-peng) of the University of Waterloo in Canada, one of the algorithm’s six creators. And in the coming years, researchers said, computer scientists will likely find ways to make it more practical and perhaps even slightly faster.

So, maybe it won't be useful directly but maybe some part are interesting for optimization of the library or for huge graphs.
